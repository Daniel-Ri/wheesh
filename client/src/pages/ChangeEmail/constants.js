export const CHANGE_EMAIL = 'ChangeEmail/CHANGE_EMAIL';
