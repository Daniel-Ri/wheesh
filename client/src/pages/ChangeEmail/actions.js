import { CHANGE_EMAIL } from './constants';

export const changeEmail = (inputs, handleSuccess, handleError) => ({
  type: CHANGE_EMAIL,
  inputs,
  handleSuccess,
  handleError,
});
