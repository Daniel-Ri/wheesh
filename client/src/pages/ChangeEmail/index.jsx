/* eslint-disable jsx-a11y/label-has-associated-control */
import { useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import { FormattedMessage, injectIntl } from 'react-intl';

import BackBtn from '@components/BackBtn';
import { useNavigate } from 'react-router-dom';
import { useState } from 'react';

import toast from 'react-hot-toast';
import { sendEmailToken } from '@containers/Client/actions';
import { Button } from '@mui/material';
import { changeEmail } from './actions';

import classes from './style.module.scss';

const ChangeEmail = ({ intl: { formatMessage } }) => {
  const dispatch = useDispatch();
  const navigate = useNavigate();

  const [inputs, setInputs] = useState({
    email: '',
    emailToken: '',
  });

  const handleInputChange = (e) => {
    setInputs((prev) => ({ ...prev, [e.target.name]: e.target.value }));
  };

  const validateEmail = () => {
    if (!inputs.email) {
      toast.error(formatMessage({ id: 'app_you_must_insert_email' }));
      return false;
    }

    if (!/^[a-zA-Z0-9.!#$%&â€™*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/.test(inputs.email)) {
      toast.error(formatMessage({ id: 'app_incorrect_email_format' }));
      return false;
    }

    return true;
  };

  const handleSuccessChange = () => {
    toast.success(formatMessage({ id: 'app_success_modify_email' }));
    navigate('/me');
  };

  const handleErrorChange = (errorMsg) => {
    toast.error(errorMsg);
  };

  const handleSubmit = () => {
    if (!validateEmail()) return;

    dispatch(changeEmail(inputs, handleSuccessChange, handleErrorChange));
  };

  const handleSuccessSendEmailToken = () => {
    toast.success(formatMessage({ id: 'app_sent_a_token_email' }));
  };

  const handleErrorSendEmailToken = (errorMsg) => {
    toast.error(errorMsg);
    setInputs((prev) => ({ ...prev, email: '' }));
  };

  const handleGenerateEmailToken = () => {
    if (!validateEmail()) return;

    const formattedInputs = { email: inputs.email, action: 'update' };
    dispatch(sendEmailToken(formattedInputs, handleSuccessSendEmailToken, handleErrorSendEmailToken));
  };

  return (
    <main className={classes.main}>
      <div className={classes.container}>
        <header>
          <BackBtn handleClickBack={() => navigate('/me')} />
          <h1>
            <FormattedMessage id="app_change_email" />
          </h1>
        </header>

        <form>
          <div className={classes.input}>
            <label htmlFor="email">E-mail</label>
            <input
              type="text"
              name="email"
              id="email"
              value={inputs.email}
              onChange={handleInputChange}
              placeholder={formatMessage({ id: 'app_please_enter_new_email' })}
              autoComplete="off"
            />
          </div>
          <div className={classes.input}>
            <input
              className={classes.verification}
              type="text"
              name="emailToken"
              id="emailToken"
              value={inputs.emailToken}
              onChange={handleInputChange}
              placeholder={formatMessage({ id: 'app_please_enter_verification_code' })}
              autoComplete="off"
            />
            <div className={classes.button}>
              <Button variant="contained" className={classes.btn} onClick={handleGenerateEmailToken}>
                <FormattedMessage id="app_generate" />
              </Button>
            </div>
          </div>
        </form>

        <div className={classes.buttons}>
          <Button variant="contained" className={classes.submit} onClick={handleSubmit}>
            <FormattedMessage id="app_submit" />
          </Button>
        </div>
      </div>
    </main>
  );
};

ChangeEmail.propTypes = {
  intl: PropTypes.object,
};

export default injectIntl(ChangeEmail);
