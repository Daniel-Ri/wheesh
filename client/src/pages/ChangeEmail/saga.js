import { takeLatest, call, put } from 'redux-saga/effects';
import { changeEmail } from '@domain/api';
import { setUser } from '@containers/Client/actions';
import { CHANGE_EMAIL } from './constants';

function* doChangeEmail({ inputs, handleSuccess, handleError }) {
  try {
    const response = yield call(changeEmail, inputs);
    yield put(setUser(response.data));
    yield call(handleSuccess);
  } catch (error) {
    yield call(handleError, error.response.data.message);
  }
}

export default function* changeEmailSaga() {
  yield takeLatest(CHANGE_EMAIL, doChangeEmail);
}
