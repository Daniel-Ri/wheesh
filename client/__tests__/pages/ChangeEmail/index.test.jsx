import { render, fireEvent } from '@utils/testHelper';
import ChangeEmail from '@pages/ChangeEmail';

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  useNavigate: () => mockNavigate,
}));

let wrapper;

beforeEach(() => {
  wrapper = render(<ChangeEmail />);
});

describe('Page ChangeEmail', () => {
  test('Renders ChangeEmail page', () => {
    const { getByTestId } = wrapper;
    expect(getByTestId('ChangeEmail')).toBeInTheDocument();
  });

  test('Should call navigate when BackBtn clicked', () => {
    const { getByTestId } = wrapper;
    const backButton = getByTestId('BackBtn');
    fireEvent.click(backButton);
    expect(mockNavigate).toHaveBeenCalledWith('/me');
  });

  test('Should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
