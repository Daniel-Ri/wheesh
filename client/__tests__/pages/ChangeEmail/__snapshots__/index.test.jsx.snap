// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page ChangeEmail Should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <main
        class="main"
        data-testid="ChangeEmail"
      >
        <div
          class="container"
        >
          <header>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium btn css-1e6y48t-MuiButtonBase-root-MuiButton-root"
              data-testid="BackBtn"
              tabindex="0"
              type="button"
            >
              Kembali
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <h1>
              Ubah Email
            </h1>
          </header>
          <form>
            <div
              class="input"
            >
              <label
                for="email"
              >
                E-mail
              </label>
              <input
                autocomplete="off"
                id="email"
                name="email"
                placeholder="Silahkan masukkan alamat email baru"
                type="text"
                value=""
              />
            </div>
            <div
              class="input"
            >
              <input
                autocomplete="off"
                class="verification"
                id="emailToken"
                name="emailToken"
                placeholder="Silakan masukkan kode verifikasi"
                type="text"
                value=""
              />
              <div
                class="button"
              >
                <button
                  class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium btn css-sghohy-MuiButtonBase-root-MuiButton-root"
                  tabindex="0"
                  type="button"
                >
                  Hasilkan
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </div>
            </div>
          </form>
          <div
            class="buttons"
          >
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium submit css-sghohy-MuiButtonBase-root-MuiButton-root"
              tabindex="0"
              type="button"
            >
              Kirim
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </main>
    </div>
  </body>,
  "container": <div>
    <main
      class="main"
      data-testid="ChangeEmail"
    >
      <div
        class="container"
      >
        <header>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium btn css-1e6y48t-MuiButtonBase-root-MuiButton-root"
            data-testid="BackBtn"
            tabindex="0"
            type="button"
          >
            Kembali
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <h1>
            Ubah Email
          </h1>
        </header>
        <form>
          <div
            class="input"
          >
            <label
              for="email"
            >
              E-mail
            </label>
            <input
              autocomplete="off"
              id="email"
              name="email"
              placeholder="Silahkan masukkan alamat email baru"
              type="text"
              value=""
            />
          </div>
          <div
            class="input"
          >
            <input
              autocomplete="off"
              class="verification"
              id="emailToken"
              name="emailToken"
              placeholder="Silakan masukkan kode verifikasi"
              type="text"
              value=""
            />
            <div
              class="button"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium btn css-sghohy-MuiButtonBase-root-MuiButton-root"
                tabindex="0"
                type="button"
              >
                Hasilkan
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </form>
        <div
          class="buttons"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium submit css-sghohy-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Kirim
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </main>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
